rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::{bindings, c_types, CStr};
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::mem::transmute;
+use core::pin::Pin;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
+        let drv_data = Box::into_raw(drv_data) as *mut c_types::c_void;
+        Ok(drv_data) as KernelResult<_>
+    };
+    // TODO don't we need to pin this?
+    let ptr = match f() {
+        Ok(ptr) => ptr,
+        Err(e) => return e.to_kernel_errno(),
+    };
+    rust_helper_platform_set_drvdata(pdev, ptr);
+    0
+}
+
+fn new_of_device_id(compatible: &CStr<'static>) -> KernelResult<bindings::of_device_id> {
+    // TODO:
+    // - fail at build time if compatible CStr doesn't fit.
+    // - can we do this safely without transmute?
+    let mut buf = [0_u8; 128];
+    if compatible.len() > buf.len() {
+        return Err(Error::EINVAL);
+    }
+    // PANIC: this will never panic: `compatible` is not longer than `buf`.
+    buf[..compatible.len()].copy_from_slice(compatible.as_bytes());
+    Ok(bindings::of_device_id {
+        // SAFETY: re-interpretation from [u8] to [i8] of same length is always safe.
+        compatible: unsafe { transmute::<[u8; 128], [i8; 128]>(buf) },
+        ..Default::default()
+    })
+}
+
+unsafe extern "C" fn remove_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let ptr = rust_helper_platform_get_drvdata(pdev);
+    let drv_data: Box<T::DrvData> = Box::from_raw(ptr as _);
+    drop(drv_data);
```

Comment:
We should consider whether we want/need to provide an explicit `remove` callback that takes ownership of the driver data. One reason this was needed in `FileOperations` was because we needed to allow `drop` to be deferred to a later time because extra work needed to be performed and we didn't want to hold up the callback; additionally, if the wrapper is an `Arc` for example, the `drop` below could just be a decrement of a refcount, and the driver wouldn't have the opportunity to run any code until it dropped to zero at a later time (in fact, the driver may _need_ to run some code for this to eventually happen).